pkg/irc/client.go

```diff
@@ -0,0 +1,252 @@
+	Address = "irc.chat.twitch.tv:6667"
+	// AddressTLS sets the address that the client will connect to in TLS mode
+	AddressTLS = "irc.chat.twitch.tv:6697"
+)
+
+// Client handles the IRC connection and incoming & outgoing messages.
+// The client requires you to respond to PING messages manually
+// as well as keep track of which channels you're connected to using the incoming JOIN & PART messages
+type Client struct {
+	user  string
+	oauth string
+
+	capabilities []string
+
+	// UseTLS determines whether the IRC connects with or without TLS, needs to be set before you call Connect, default = true
+	UseTLS bool
+
+	read  chan string
+	write chan []byte
+
+	serverDisconnect closer
+	clientDisconnect closer
+
+	onConnect func()
+	onMessage func(msg *Message, err error)
+}
+
+// New returns a new client
+func New(user, oauth string) *Client {
+	return &Client{
+		user:   user,
+		oauth:  oauth,
+		UseTLS: true,
+		read:   make(chan string, ReadBuffer),
+		write:  make(chan []byte, WriteBuffer),
+	}
+}
+
+// NewAnon returns an anonymous client, useful for testing, or small read-only bots
+func NewAnon() *Client {
+	return &Client{
+		user:   "justinfan4321",
+		oauth:  "oauth",
+		UseTLS: true,
+		read:   make(chan string, ReadBuffer),
+		write:  make(chan []byte, WriteBuffer),
+	}
+}
+
+// WithCapabilities adds twitch-irc specific capabilities to a New client, use the constants defined in capabilities.go
+func (c *Client) WithCapabilities(caps ...string) *Client {
+	c.capabilities = caps
+	return c
+}
+
+// Connect starts the IRC connection
+func (c *Client) Connect() (err error) {
+	dialer := &net.Dialer{KeepAlive: time.Second * 10}
+	var conn net.Conn
+	if c.UseTLS {
+		conn, err = tls.DialWithDialer(dialer, "tcp", AddressTLS, &tls.Config{MinVersion: tls.VersionTLS12})
+	} else {
+		conn, err = dialer.Dial("tcp", Address)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	c.clientDisconnect.reset()
+	c.serverDisconnect.reset()
+
+	err = c.requestCapabilities(conn)
+	if err != nil {
+		return err
+	}
+
+	err = c.login(conn)
+	if err != nil {
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go c.startReader(conn, &wg)
+	go c.startWriter(conn, &wg)
+
+	// blocks here, until server disconnects or c.Disconnect() is called,
+	//the error we get from here tells us whether the client disconnected, or the server disconnected us
+	err = c.startHandler()
+
+	// close all channels & connections to make sure there's no memory leaks
+	conn.Close()
+	c.serverDisconnect.close()
+	c.clientDisconnect.close()
+
+	// Wait until both reader & writer are closed
+	wg.Wait()
+
+	return
+}
+
+// Disconnect closes the IRC connection
+func (c *Client) Disconnect() {
+	c.clientDisconnect.close()
+}
+
+// Join makes the client join the passed channels
+func (c *Client) Join(channels ...string) {
+	c.sendString("JOIN " + appendChannels(channels...))
+}
+
+// Depart makes the client leave the passed channels
+func (c *Client) Depart(channels ...string) {
```

Comment:
nit, this could just be `Part()`
```suggestion
func (c *Client) Part(channels ...string) {
```

Reply:
Done